lib/Table.js

```diff
@@ -75,277 +74,307 @@ const compareIndexes = function compareIndexes(local, remote) {
-        if (gsi.IndexStatus !== 'ACTIVE') {
-          active = false;
-        }
-      });
-      if (!active) {
-        debug('Waiting for Active again - %s', table.name);
-        setTimeout(waitForActive, 500);
-      } else {
-        table.active = true;
-        deferred.resolve();
+
+    let table = this;
+
+    let timeoutAt = Date.now() + timeout;
+
+    function waitForActive() {
+      debug('Waiting...');
+      if (Date.now() > timeoutAt) {
+        return reject(
+          new Error('Wait for Active timed out after ' + timeout + ' ms.')
+        );
       }
-    })
-    .catch(function (err) {
-      if (err && err.code === 'ResourceNotFoundException') {
+      if (!table.initialized) {
         return setTimeout(waitForActive, 10);
       }
-      debug('Error waiting for active', err.stack);
-      return deferred.reject(err);
-    });
-  }
-
-  waitForActive();
+      table.describe()
+        .then(function (data) {
+          let active = (data.Table.TableStatus === 'ACTIVE');
+          let indexes = data.Table.GlobalSecondaryIndexes || [];
+          indexes.forEach(function (gsi) {
+            //debug('waitForActive Index Check: %s', JSON.stringify(gsi, null, 2));
+            debug('index %s.IndexStatus is %s', gsi.IndexName, gsi.IndexStatus);
+            if (gsi.IndexStatus !== 'ACTIVE') {
+              active = false;
+            }
+          });
+          if (!active) {
+            debug('Waiting for Active again - %s', table.name);
+            setTimeout(waitForActive, 500);
+          } else {
+            table.active = true;
+            resolve();
+          }
+        })
+        .catch(function (err) {
+          if (err && err.code === 'ResourceNotFoundException') {
+            return setTimeout(waitForActive, 10);
+          }
+          debug('Error waiting for active', err.stack);
+          return reject(err);
+        });
+    }
 
-  return deferred.promise.nodeify(next);
+    waitForActive();
+  })
 };
 
 Table.prototype.describeTTL = function (next) {
-  debug('Describing ttl for table, %s', this.name);
-  const deferred = Q.defer();
 
-  let ddb = this.base.ddb();
+  if(next){
+    this.describeTTL()
+      .then(result => {next(undefined,result)}, err => {next(err)})
+    return
+  }
 
-  let params = {
-    TableName: this.name
-  };
-  ddb.describeTimeToLive(params, function(err, ttlDescription) {
-    if (err) {
-      return deferred.reject(err);
-    }
+  return new Promise((resolve, reject) => {
+    debug('Describing ttl for table, %s', this.name);
+
+    let ddb = this.base.ddb();
 
-    return deferred.resolve(ttlDescription);
-  });
+    let params = {
+      TableName: this.name
+    };
+    ddb.describeTimeToLive(params, function(err, ttlDescription) {
+      if (err) {
+        return reject(err);
+      }
 
-  return deferred.promise.nodeify(next);
+      return resolve(ttlDescription);
+    });
+
+    return promise.nodeify(next);
```

Comment:
Where is `promise` defined? And why are we using two different methods for handling the callbacks? We do a lot of the following:

```
this.describeTTL()
       .then(result => {next(undefined,result)}, err => {next(err)})
return
```

This kinda ties into my question at the very top. And maybe if this thing exists and works, we could replace a lot of that code with this better version.